File: modules/platforms/php/src/Apache/Ignite/Impl/Binary/BinarySchema.php
<?php
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace Apache\Ignite\Impl\Binary;

class BinarySchema
{
    /** FNV1 hash offset basis. */
    const FNV1_OFFSET_BASIS = 0x811C9DC5;
    /** FNV1 hash prime. */
    const FNV1_PRIME = 0x01000193;
    
    private $id;
    private $fieldIds;
    private $isValid;
    
    public function __construct()
    {
        $this->id = BinarySchema::FNV1_OFFSET_BASIS;
        $this->fieldIds = [];
        $this->isValid = true;
    }
    
    public function getId(): int
    {
        return $this->id;
    }
    
    public function getFieldIds(): array
    {
        return array_keys($this->fieldIds);
    }

    public function finalize(): void
    {
        if (!$this->isValid) {
            $this->id = BinarySchema::FNV1_OFFSET_BASIS;
            foreach ($this->fieldIds as $key => $value) {
                $this->id = BinarySchema::updateSchemaId($this->id, $key);
            }
            $this->isValid = true;
        }
    }

    public function cloneSchema(): BinarySchema
    {
        $result = new BinarySchema();
        $result->id = $this->id;
        $result->fieldIds = array($this->fieldIds); new Set(this._fieldIds);
        $result->isValid = $this->isValid;
        return result;
    }
    
    public function addField(int $fieldId): void
    {
        if (!$this->hasField($fieldId)) {
            $this->fieldIds[$fieldId] = true;
            if ($this->isValid) {
                $this->updateSchemaId($fieldId);
            }
        }
    }

    public function removeField(int $fieldId): void
    {
        if ($this->hasField($fieldId)) {
            unset($this->fieldIds[$fieldId]);
            $this->isValid = false;
        }
    }

    public function hasField(int $fieldId): bool
    {
        return array_key_exists($fieldId, $this->fieldIds);
    }

    private function updateSchemaId(int $fieldId): void
    {
        $this->id = $this->id ^ ($fieldId & 0xFF);
        $this->id = $this->id * BinarySchema::FNV1_PRIME;
        $this->id &= 0xFFFFFFFF; // Convert to 32bit integer
        $this->id = $this->id ^ (($fieldId >> 8) & 0xFF);
        $this->id = $this->id * BinarySchema::FNV1_PRIME;
        $this->id &= 0xFFFFFFFF; // Convert to 32bit integer
        $this->id = $this->id ^ (($fieldId >> 16) & 0xFF);
        $this->id = $this->id * BinarySchema::FNV1_PRIME;
        $this->id &= 0xFFFFFFFF; // Convert to 32bit integer
        $this->id = $this->id ^ (($fieldId >> 24) & 0xFF);
        $this->id = $this->id * BinarySchema::FNV1_PRIME;
        $this->id &= 0xFFFFFFFF; // Convert to 32bit integer
        $this->id = BinaryUtils::intval32($this->id);
    }

    public function write(MessageBuffer $buffer): void
    {
        $this->finalize();
        // schema id
        $buffer->writeInteger($this->id);
        // fields count
        $buffer->writeInteger(count($this->fieldIds));
        // field ids
        foreach ($this->fieldIds as $key => $value) {
            $buffer->writeInteger($key);
        }
    }
    
    public function read(MessageBuffer $buffer): void
    {
        // schema id
        $this->id = $buffer->readInteger();
        // fields count
        $fieldsCount = $buffer->readInteger();
        // field ids
        for ($i = 0; $i < $fieldsCount; $i++) {
            $this->fieldIds[$buffer->readInteger()] = true;
        }
    }
}

File: modules/core/src/main/java/org/apache/ignite/internal/binary/BinarySchema.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.ignite.internal.binary;

import java.io.DataInput;
import java.io.DataOutput;
import java.io.IOException;

/**
 * Schema describing binary object content.
 */
public interface BinarySchema {
    /**
     * @return Schema ID.
     */
    public int schemaId();

    /**
     * Try speculatively confirming order for the given field name.
     *
     * @param expOrder Expected order.
     * @param expName Expected name.
     * @return Field ID.
     */
    public Confirmation confirmOrder(int expOrder, String expName);

    /**
     * Add field name.
     *
     * @param order Order.
     * @param name Name.
     */
    public void clarifyFieldName(int order, String name);

    /**
     * Get field ID by order in footer.
     *
     * @param order Order.
     * @return Field ID.
     */
    public int fieldId(int order);

    /**
     * Get field order in footer by field ID.
     *
     * @param id Field ID.
     * @return Offset or {@code 0} if there is no such field.
     */
    public int order(int id);

    /**
     * The object implements the writeTo method to save its contents
     * by calling the methods of DataOutput for its primitive values and strings or
     * calling the writeTo method for other objects.
     *
     * @param out the stream to write the object to.
     * @throws IOException Includes any I/O exceptions that may occur.
     */
    public void writeTo(DataOutput out) throws IOException;

    /**
     * The object implements the readFrom method to restore its
     * contents by calling the methods of DataInput for primitive
     * types and strings or calling readFrom for other objects.  The
     * readFrom method must read the values in the same sequence
     * and with the same types as were written by writeTo.
     *
     * @param in the stream to read data from in order to restore the object
     * @throws IOException if I/O errors occur
     */
    public void readFrom(DataInput in) throws IOException;

    /**
     * Gets field ids array.
     *
     * @return Field ids.
     */
    public int[] fieldIds();

    /**
     * Order confirmation result.
     */
    public enum Confirmation {
        /** Confirmed. */
        CONFIRMED,

        /** Denied. */
        REJECTED,

        /** Field name clarification is needed. */
        CLARIFY
    }
}
